In [ ]:
!pip install contractions

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.9/289.9 kB 11.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.7/110.7 kB 6.9 MB/s eta 0:00:00


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
!kaggle datasets download -d nayan082/sentimentnewdataset

Dataset URL: https://www.kaggle.com/datasets/nayan082/sentimentnewdataset
License(s): copyright-authors
  0% 0.00/568k [00:00<?, ?B/s]
100% 568k/568k [00:00<00:00, 131MB/s]


In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
import contractions
nltk.download('stopwords')


def preprocess_text(text):
    text = text.lower()
    text = contractions.fix(text)
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)
    text = re.sub(r'@\w+', '', text)
    text = re.sub(r'#(\w+)', r'\1', text)
    text = re.sub(r'[^\w\s\U0001F600-\U0001F64F\U0001F300-\U0001F5FF\U0001F680-\U0001F6FF\U0001F700-\U0001F77F]', '', text)
    text = re.sub(r'\b\w\b', '', text)
    text = re.sub(r'(.)\1+', r'\1\1', text)
    stop_words = set(stopwords.words('english'))
    text = ' '.join([word for word in text.split() if word not in stop_words])
    text = re.sub(r'\s+', ' ', text).strip()

    return text


# df = pd.read_csv('/content/drive/MyDrive/140.csv',encoding='ISO-8859-1')
# df['text'] = df.iloc[:, -1]
# df['processed_text'] = df['text'].apply(preprocess_text)
# df.to_csv('/content/drive/MyDrive/preprocessed_dataset.csv', index=False)
# print(df[['text', 'processed_text']].head())
# df['sentiment'] = df.iloc[:,0]
# print(df.head())
# print(df.tail())


emoji_df = pd.read_csv('/content/drive/MyDrive/large_tweets_data.csv')
emoji_df['processed_text'] = emoji_df['Text'].apply(preprocess_text)
emoji_df = emoji_df[['Text', 'Sentiment','processed_text']]
df_140 = pd.read_csv('/content/drive/MyDrive/reduced_140data.csv',encoding='ISO-8859-1')
df_140 = df_140[['Text','Sentiment','processed_text']]
combined_df = pd.concat([emoji_df, df_140], ignore_index=True)
combined_df.to_csv('/content/drive/MyDrive/preprocessed_dataset2.csv', index=False)

#emoji_df.to_csv('/content/drive/MyDrive/preprocessed_dataset2.csv', index=False)
print(combined_df.head())
print(combined_df.tail())



[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


                                                Text  Sentiment  \
0  @honkhaikey this one girl has cat maid q as th...          0   
1                          So embarrassing. Lmfao. 😭          0   
2  @BibIeCollege &amp; all the coloured people re...          0   
3  @Sheer__Opulence Small Guards under 6feet...He...          0   
4  @miIflvr I’m actually 5’9 just a cm off of 5’10 😭          0   

                            processed_text  
0                  one girl cat maid pfp 😭  
1                     embarrassing lmfao 😭  
2  amp coloured people reading kerimelle 😭  
3      small guards 6feethere way get 65 😭  
4                     actually 59 cm 510 😭  
                                                     Text  Sentiment  \
316006  @Issascrazyworld I was trying to win some of t...          0   
316007  Lesson learned: Tennis, table tennis and badmi...          3   
316008  Picking up my photos from my photo shoot this ...          3   
316009                      is drown in

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

df_140 = pd.read_csv('/content/drive/MyDrive/140data.csv', encoding='ISO-8859-1')
df_140['Text'] = df_140.iloc[:, -1]
df_140['Sentiment'] = df_140.iloc[:, 0]
df_140 = df_140.dropna(subset=['Sentiment'])


df_140['Sentiment'] = df_140['Sentiment'].replace({4: 3})

df_sampled, _ = train_test_split(df_140, train_size=300000, stratify=df_140['Sentiment'], random_state=42)
df_sampled['processed_text'] = df_sampled['Text'].apply(preprocess_text)
df_sampled.to_csv('/content/drive/MyDrive/reduced_140data.csv', index=False)
print(df_sampled['Sentiment'].value_counts())
print(df_sampled.head())



ParserError: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.

In [ ]:
!pip install emoji
from sklearn.model_selection import train_test_split
import emoji
import pandas as pd
import re
import nltk

df2 = pd.read_csv('/content/drive/MyDrive/preprocessed_dataset2.csv')
print(df2.head())
X_train, X_test, y_train, y_test = train_test_split(df2['processed_text'], df2['Sentiment'], test_size=0.2, random_state=42)

print(f"Training set size: {len(X_train)}, Testing set size: {len(X_test)}")


def replace_emojis(text):
    if isinstance(text, str):
        return emoji.demojize(text, delimiters=(" ", " "))
    else:
        return text

df2['text_with_emojis_replaced'] = df2['processed_text'].apply(replace_emojis)

print(df2.head())

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 586.9/586.9 kB 11.5 MB/s eta 0:00:00
                                                Text  Sentiment  \
0  @honkhaikey this one girl has cat maid q as th...          0   
1                          So embarrassing. Lmfao. 😭          0   
2  @BibIeCollege &amp; all the coloured people re...          0   
3  @Sheer__Opulence Small Guards under 6feet...He...          0   
4  @miIflvr I’m actually 5’9 just a cm off of 5’10 😭          0   

                            processed_text  
0                  one girl cat maid pfp 😭  
1                     embarrassing lmfao 😭  
2  amp coloured people reading kerimelle 😭  
3      small guards 6feethere way get 65 😭  
4                     actually 59 cm 510 😭  
Training set size: 252808, Testing set size: 63203
                                                Text  Sentiment  \
0  @honkhaikey this one girl has cat maid q as th...          0   
1                          So embarrassing. Lmfao. 😭         

In [ ]:
import nltk
from nltk.corpus import sentiwordnet as swn
from nltk.corpus import wordnet as wn

nltk.download('sentiwordnet')
nltk.download('wordnet')

positive_words = set()
negative_words = set()

all_words = set(wn.all_lemma_names())

for word in all_words:

    senti_synsets = list(swn.senti_synsets(word))

    if senti_synsets:

        for senti_synset in senti_synsets:

            positive_score = senti_synset.pos_score()
            negative_score = senti_synset.neg_score()

            if positive_score > 0.5:
                positive_words.add(word)
            elif negative_score > 0.5:
                negative_words.add(word)

print("Positive Words:", positive_words)
print("Negative Words:", negative_words)

[nltk_data] Downloading package sentiwordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/sentiwordnet.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


Positive Words: {'reassertion', 'infectious_agent', 'honeyed_words', 'spiritedness', 'commonsensible', 'meanspirited', 'astounded', 'venerating', 'well-situated', 'humane', 'loved', 'sensationalism', 'publishable', 'provocative', 'starry-eyed', 'puritanical', 'white_magic', 'plucky', 'revered', 'conscious', 'sportiveness', 'pleased', 'mom', 'pall', 'intelligent', 'playfulness', 'verisimilitude', 'convivial', 'white_elephant', 'probity', 'noticeable', 'good-humored', 'awing', 'seemly', 'corrective', 'respond', 'diviner', 'healthful', 'distinction', 'relapsing', 'discoverable', 'belong', 'euphonic', 'marvelous', 'anthelmintic', 'modishness', 'asteriated', 'rightness', 'caroche', 'savory', 'rarified', 'rosy-cheeked', 'sneak', 'appetent', 'direct', 'isolable', 'attractive', 'elfin', 'essential', 'aesthetical', 'good_humor', 'banner', 'lordly', 'trustful', 'truth', 'going-over', 'provable', 'authentic', 'certified', 'cleanable', 'amusing', 'strike', 'perked_up', 'clean_bill_of_health', 'cha

In [ ]:
!pip install emoji
import emoji
import re
import pandas as pd


positive_emoji_list = [
    '😊', '😂', '❤️', '😍', '🥰', '😄', '😃', '😆', '😇',
    '✨', '🌈', '🥳', '🎉', '🎊', '😌', '🤗', '🌼', '💖',
    '😸', '🤩', '😺', '🙌', '👍', '🤝', '🙏', '🎶', '🍀',
    '💫', '🌟', '☀️', '🌹', '🥂', '💞', '🍓', '🦋', '🍉',
    '😘', '😻', '😋', '🥇', '🏆', '🍧', '🍰', '🍦', '🌸',
    '🌻', '🐶', '🐱', '🐰', '🚀', '🌺', '🌙', '🔆', '💗',
    '💪', '🎈', '🦄', '😇', '🕊️', '🍕', '🤩', '🧁', '💖',
    '🦩', '🎀', '🌻', '🌞', '🐼', '🧸', '🍕', '🤗', '🪴', '🎉'
]


negative_emoji_list = [
    '😢', '😡', '😠', '💔', '😔', '😩', '😣', '😖', '😤',
    '😱', '😵', '😷', '🤬', '👎', '💀', '☠️', '🖤', '🩸',
    '🌧️', '⚡', '😓', '😥', '🥀', '😒', '😞', '🤯', '😿',
    '😔', '🚫', '💔', '😭', '😰', '😳', '😨', '😬', '😩',
    '😹', '👺', '🧟‍♂️', '🤮', '😕', '🤢', '🥵', '😤', '💩',
    '👿', '🖤', '🩸', '🥀', '😫', '😰', '😵‍💫', '🔪', '🧨',
    '🤕', '🧟‍♀️', '😨', '😳', '😞', '💩', '💔', '☠️', '🔪'
]


strong_positive_keywords = positive_words #['love', 'amazing', 'fantastic', 'great', 'happy', 'joy', 'wonderful', 'best', 'like', 'enjoy']
strong_negative_keywords = negative_words #['hate', 'awful', 'terrible', 'bad', 'worst', 'sad', 'angry', 'dislike', 'despise', 'loathe']

def calculate_emoji_score(text):
    if isinstance(text, str):
        score = 0
        emojis_in_text = [char for char in text if char in emoji.EMOJI_DATA]
        text_lower = text.lower()
        for keyword in strong_positive_keywords:
            if keyword in text_lower:
                score += 2
        for keyword in strong_negative_keywords:
            if keyword in text_lower:
                score -= 2

        for e in emojis_in_text:
            if e in positive_emoji_list:
                score += 1
            elif e in negative_emoji_list:
                score -= 1

        return score


df2['emoji_score'] = df2['processed_text'].apply(calculate_emoji_score)

# Print sample results
print(df2[['processed_text', 'emoji_score']].head())
print(df2[['processed_text', 'emoji_score']].tail())


                            processed_text  emoji_score
0                  one girl cat maid pfp 😭          1.0
1                     embarrassing lmfao 😭         -3.0
2  amp coloured people reading kerimelle 😭         -1.0
3      small guards 6feethere way get 65 😭         -3.0
4                     actually 59 cm 510 😭         -1.0
                                           processed_text  emoji_score
316006              trying win cash know get squat though         -2.0
316007  lesson learned tennis table tennis badminton u...          0.0
316008               picking photos photo shoot wednesday         -2.0
316009                                drown naked revenge          0.0
316010      new blog post working miniminscom weight loss         -4.0


In [ ]:
import numpy as np

def load_emoji2vec(embedding_file):
    emoji_embeddings = {}
    with open(embedding_file, 'r', encoding='utf-8') as f:
        for line in f:
            values = line.split()
            emoji = values[0]
            vector = np.asarray(values[1:], "float32")
            emoji_embeddings[emoji] = vector
    return emoji_embeddings

emoji2vec = load_emoji2vec('/content/drive/MyDrive/emoji2vec.txt')

def get_emoji_embeddings(text):
  if isinstance(text, str):
    embeddings = []
    for char in text:
        if char in emoji2vec:
            embeddings.append(emoji2vec[char])
    if embeddings:
        return np.mean(embeddings, axis=0)
    else:
        return np.zeros(300)

df2['emoji_embeddings'] = df2['processed_text'].apply(get_emoji_embeddings)
print(df2.head())
print(df2.tail())

                                                Text  Sentiment  \
0  @honkhaikey this one girl has cat maid q as th...          0   
1                          So embarrassing. Lmfao. 😭          0   
2  @BibIeCollege &amp; all the coloured people re...          0   
3  @Sheer__Opulence Small Guards under 6feet...He...          0   
4  @miIflvr I’m actually 5’9 just a cm off of 5’10 😭          0   

                            processed_text  \
0                  one girl cat maid pfp 😭   
1                     embarrassing lmfao 😭   
2  amp coloured people reading kerimelle 😭   
3      small guards 6feethere way get 65 😭   
4                     actually 59 cm 510 😭   

                           text_with_emojis_replaced  emoji_score  \
0         one girl cat maid pfp  loudly_crying_face           1.0   
1            embarrassing lmfao  loudly_crying_face          -3.0   
2  amp coloured people reading kerimelle  loudly_...         -1.0   
3  small guards 6feethere way get 65  loudly

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
import numpy as np


df2 = df2.dropna(subset=['text_with_emojis_replaced'])

vectorizer = TfidfVectorizer(max_features=5000)
X_text = vectorizer.fit_transform(df2['text_with_emojis_replaced'])

# Convert emoji embeddings to a matrix format
X_emoji = np.vstack(df2['emoji_embeddings'].values)

# Combine text vectors and emoji embeddings
X_combined = np.hstack([X_text.toarray(), X_emoji])

y = df2['Sentiment'].values
X_train, X_test, y_train, y_test = train_test_split(X_combined, y, test_size=0.2, random_state=42)
print(X_text)





In [ ]:
from sklearn.naive_bayes import BernoulliNB
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score

bnb = BernoulliNB()
svm = SVC()
lr = LogisticRegression(max_iter=1000)


bnb.fit(X_train, y_train)
bnb_preds = bnb.predict(X_test)
print(f"BNB Accuracy: {accuracy_score(y_test, bnb_preds)}, F1: {f1_score(y_test, bnb_preds, average='weighted')}")

svm.fit(X_train, y_train)
svm_preds = svm.predict(X_test)
print(f"SVM Accuracy: {accuracy_score(y_test, svm_preds)}, F1: {f1_score(y_test, svm_preds, average='weighted')}")

lr.fit(X_train, y_train)
lr_preds = lr.predict(X_test)
test=""
print(f"LR Accuracy: {accuracy_score(y_test, lr_preds)}, F1: {f1_score(y_test, lr_preds, average='weighted')}")


BNB Accuracy: 0.7202023395510592, F1: 0.7256976517504977


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000)  # Set the max features as desired
X_text = vectorizer.fit_transform(df2['text_with_emojis_replaced'])

# Get vocabulary size
vocab_size = len(vectorizer.vocabulary_)

# Define max sequence length
max_seq_length = 100

from tensorflow.keras.preprocessing.sequence import pad_sequences


X_train_padded = pad_sequences(X_train, maxlen=max_seq_length)
X_test_padded = pad_sequences(X_test, maxlen=max_seq_length)


In [ ]:
import tensorflow as tf
  from tensorflow.keras.layers import Embedding, LSTM, Bidirectional, Conv1D, MaxPooling1D, Dense, Dropout, Flatten
  from tensorflow.keras.models import Sequential
  from sklearn.metrics import accuracy_score, f1_score


  num_classes = len(np.unique(y_train))

  # Building the BiLSTM-CNN Model for multiclass classification
  model = Sequential([
      Embedding(input_dim=vocab_size, output_dim=300),
      Bidirectional(LSTM(128, return_sequences=True)),
      Conv1D(64, kernel_size=5, activation='relu'),
      MaxPooling1D(pool_size=2),
      Flatten(),
      Dense(64, activation='relu'),
      Dropout(0.5),
      Dense(num_classes, activation='softmax')  # Use softmax for multiclass classification
  ])

  # Compile the model with categorical crossentropy loss
  model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])


  # Training the model
  model.fit(X_train_padded, y_train, validation_data=(X_test_padded, y_test), epochs=5, batch_size=32)

  # Evaluating the model
  dl_preds = model.predict(X_test_padded)
  dl_preds = np.argmax(dl_preds, axis=1)  # Convert probabilities to class labels
  print(f"DL Model Accuracy: {accuracy_score(y_test, dl_preds)}, F1: {f1_score(y_test, dl_preds, average='weighted')}")



Epoch 1/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 232s 571ms/step - accuracy: 0.6146 - loss: 0.8405 - val_accuracy: 0.7126 - val_loss: 0.6446
Epoch 2/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 261s 570ms/step - accuracy: 0.7017 - loss: 0.6417 - val_accuracy: 0.7161 - val_loss: 0.6177
Epoch 3/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 248s 627ms/step - accuracy: 0.7001 - loss: 0.6417 - val_accuracy: 0.7167 - val_loss: 0.6218
Epoch 4/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 270s 648ms/step - accuracy: 0.7066 - loss: 0.6266 - val_accuracy: 0.7148 - val_loss: 0.6342
Epoch 5/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 228s 577ms/step - accuracy: 0.7153 - loss: 0.6320 - val_accuracy: 0.7189 - val_loss: 0.6042
99/99 ━━━━━━━━━━━━━━━━━━━━ 13s 129ms/step
DL Model Accuracy: 0.7189377173569396, F1: 0.7270869369871323


In [ ]:
import numpy as np
import tensorflow as tf
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Embedding, LSTM, Bidirectional, Conv1D, MaxPooling1D, Dense, Dropout, Flatten, Concatenate, Input
from tensorflow.keras.models import Model
from sklearn.metrics import accuracy_score, f1_score


vectorizer = TfidfVectorizer(max_features=5000)
X_text = vectorizer.fit_transform(df2['text_with_emojis_replaced'])


vocab_size = len(vectorizer.vocabulary_)


max_seq_length = 100

X_text_array = X_text.toarray()
X_text_padded = pad_sequences(X_text_array, maxlen=max_seq_length, padding='post')
X_emoji_embeddings = np.vstack(df2['emoji_embeddings'].values)
X_emoji_scores = df2['emoji_score'].values
X_combined = np.hstack([X_text_padded, X_emoji_embeddings])
y = df2['Sentiment'].values


X_train_combined, X_test_combined, y_train, y_test = train_test_split(X_combined, y, test_size=0.2, random_state=42)
X_train_padded = X_train_combined[:, :max_seq_length]
X_train_emoji_embeddings = X_train_combined[:, max_seq_length:max_seq_length+300]
X_train_emoji_scores = X_train_combined[:, -1]

X_test_padded = X_test_combined[:, :max_seq_length]
X_test_emoji_embeddings = X_test_combined[:, max_seq_length:max_seq_length+300]
X_test_emoji_scores = X_test_combined[:, -1]  g

num_classes = len(np.unique(y_train))

text_input = Input(shape=(max_seq_length,), name='text_input')
emoji_input = Input(shape=(300,), name='emoji_input')
emoji_score_input = Input(shape=(1,), name='emoji_score_input')

embedding_layer = Embedding(input_dim=vocab_size, output_dim=300)(text_input)  # Use the vocabulary size from TF-IDF
bilstm = Bidirectional(LSTM(128, return_sequences=True))(embedding_layer)
conv1d = Conv1D(64, kernel_size=5, activation='relu')(bilstm)
maxpool = MaxPooling1D(pool_size=2)(conv1d)
flatten = Flatten()(maxpool)


combined_features = Concatenate()([flatten, emoji_input, emoji_score_input])


dense1 = Dense(64, activation='relu')(combined_features)
dropout = Dropout(0.5)(dense1)
output = Dense(num_classes, activation='softmax')(dropout)  # For multiclass classification

model = Model(inputs=[text_input, emoji_input, emoji_score_input], outputs=output)

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.summary()
model.fit([X_train_padded, X_train_emoji_embeddings, X_train_emoji_scores], y_train,
          validation_data=([X_test_padded, X_test_emoji_embeddings, X_test_emoji_scores], y_test),
          epochs=5, batch_size=32)

dl_preds = model.predict([X_test_padded, X_test_emoji_embeddings, X_test_emoji_scores])
dl_preds = np.argmax(dl_preds, axis=1)  # Convert probabilities to class labels


print(f"DL Model Accuracy: {accuracy_score(y_test, dl_preds)}, F1: {f1_score(y_test, dl_preds, average='weighted')}")




Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ text_input (InputLayer)   │ (None, 100)            │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ embedding (Embedding)     │ (None, 100, 300)       │      1,500,000 │ text_input[0][0]       │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ bidirectional             │ (None, 100, 256)       │        439,296 │ embedding[0][0]        │
│ (Bidirectional)           │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv1d (Conv1D)           │ (None, 96, 64)         │         81,984 │ bidirectional[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling1d             │ (None, 48, 64)         │              0 │ conv1d[0][0]           │
│ (MaxPooling1D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ flatten (Flatten)         │ (None, 3072)           │              0 │ max_pooling1d[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ emoji_input (InputLayer)  │ (None, 300)            │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ emoji_score_input         │ (None, 1)              │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ concatenate (Concatenate) │ (None, 3373)           │              0 │ flatten[0][0],         │
│                           │                        │                │ emoji_input[0][0],     │
│                           │                        │                │ emoji_score_input[0][… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense (Dense)             │ (None, 64)             │        215,936 │ concatenate[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dropout (Dropout)         │ (None, 64)             │              0 │ dense[0][0]            │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_1 (Dense)           │ (None, 4)              │            260 │ dropout[0][0]          │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 2,237,476 (8.54 MB)

 Trainable params: 2,237,476 (8.54 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 227s 552ms/step - accuracy: 0.6986 - loss: 0.6853 - val_accuracy: 0.7215 - val_loss: 0.6223
Epoch 2/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 213s 539ms/step - accuracy: 0.7100 - loss: 0.6214 - val_accuracy: 0.7227 - val_loss: 0.6048
Epoch 3/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 214s 541ms/step - accuracy: 0.7131 - loss: 0.6246 - val_accuracy: 0.7199 - val_loss: 0.6029
Epoch 4/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 269s 559ms/step - accuracy: 0.7239 - loss: 0.6086 - val_accuracy: 0.7215 - val_loss: 0.5999
Epoch 5/5
396/396 ━━━━━━━━━━━━━━━━━━━━ 263s 562ms/step - accuracy: 0.7223 - loss: 0.5939 - val_accuracy: 0.7237 - val_loss: 0.6028
99/99 ━━━━━━━━━━━━━━━━━━━━ 14s 135ms/step
DL Model Accuracy: 0.7236800505848877, F1: 0.731964406596462


In [ ]:
# using glove and experimenting more with the model
import kagglehub
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Embedding, LSTM, Bidirectional, Conv1D, MaxPooling1D, Dense, Dropout, Flatten, Concatenate, Input, BatchNormalization, Attention
from tensorflow.keras.models import Model
from sklearn.metrics import accuracy_score, f1_score

path = kagglehub.dataset_download("rtatman/glove-global-vectors-for-word-representation")
print("Path to dataset files:", path)


def load_glove_embeddings(glove_path, embedding_dim):
    embeddings_index = {}
    with open(glove_path, 'r', encoding='utf-8') as f:
        for line in f:
            values = line.split()
            word = values[0]
            coefs = np.asarray(values[1:], dtype='float32')
            embeddings_index[word] = coefs
    return embeddings_index


glove_path = '/root/.cache/kagglehub/datasets/rtatman/glove-global-vectors-for-word-representation/versions/1/glove.6B.200d.txt'
embedding_dim = 200
glove_embeddings = load_glove_embeddings(glove_path, embedding_dim)

# We are tokenizing here, previously we used TF-IDF vectors
max_num_words = 5000
max_seq_length = 100
def fix_embedding_size(embedding, expected_size=300):
    if isinstance(embedding, str):  # If it's a string representation
        try:
            embedding = np.array(ast.literal_eval(embedding), dtype=float)  # Safely convert string to array
        except (ValueError, SyntaxError):
            embedding = np.zeros(expected_size)  # If conversion fails, use zeros
    elif not isinstance(embedding, np.ndarray) or embedding.size != expected_size:
        embedding = np.zeros(expected_size)  # Default to zero if not correct size
    return embedding

df2['emoji_embeddings'] = df2['emoji_embeddings'].apply(lambda x: fix_embedding_size(x, expected_size=300))
df2['text_with_emojis_replaced'] = df2['text_with_emojis_replaced'].fillna('')  # Replace NaNs with empty strings

# Ensure all values are strings
df2['text_with_emojis_replaced'] = df2['text_with_emojis_replaced'].astype(str)
print(df2['text_with_emojis_replaced'].tail())
print(df2['emoji_embeddings'])

tokenizer = Tokenizer(num_words=max_num_words)
tokenizer.fit_on_texts(df2['text_with_emojis_replaced'])
X_text_sequences = tokenizer.texts_to_sequences(df2['text_with_emojis_replaced'])
X_text_padded = pad_sequences(X_text_sequences, maxlen=max_seq_length)

# Creating the embedding matrix
word_index = tokenizer.word_index
vocab_size = min(max_num_words, len(word_index)) + 1

embedding_matrix = np.zeros((vocab_size, embedding_dim))
for word, i in word_index.items():
    if i >= max_num_words:
        continue
    embedding_vector = glove_embeddings.get(word)
    if embedding_vector is not None:
        embedding_matrix[i] = embedding_vector


X_emoji_embeddings = np.vstack(df2['emoji_embeddings'].values)
X_emoji_scores = df2['emoji_score'].values


X_combined = np.hstack([X_text_padded, X_emoji_embeddings])
y = df2['Sentiment'].values

X_train_combined, X_test_combined, y_train, y_test = train_test_split(X_combined, y, test_size=0.2, random_state=42)

X_train_padded = X_train_combined[:, :max_seq_length]
X_train_emoji_embeddings = X_train_combined[:, max_seq_length:max_seq_length+300]
X_train_emoji_scores = X_train_combined[:, -1]

X_test_padded = X_test_combined[:, :max_seq_length]
X_test_emoji_embeddings = X_test_combined[:, max_seq_length:max_seq_length+300]
X_test_emoji_scores = X_test_combined[:, -1]



num_classes = len(np.unique(y_train))


text_input = Input(shape=(max_seq_length,), name='text_input')
emoji_input = Input(shape=(300,), name='emoji_input')
emoji_score_input = Input(shape=(1,), name='emoji_score_input')

# GloVe Embedding layer
embedding_layer = Embedding(input_dim=vocab_size,
                            output_dim=embedding_dim,
                            weights=[embedding_matrix],
                            input_length=max_seq_length,
                            trainable=False)(text_input)

# using LSTM + Conv1D
bilstm = Bidirectional(LSTM(128, return_sequences=True))(embedding_layer)
conv1d = Conv1D(64, kernel_size=5, activation='relu')(bilstm)
maxpool = MaxPooling1D(pool_size=2)(conv1d)

# Attention mechanism
attention = Attention()([maxpool, maxpool])
flatten = Flatten()(attention)

# Combine text and emoji features
combined_features = Concatenate()([flatten, emoji_input, emoji_score_input])

# Dense layers
dense1 = Dense(128, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(0.01))(combined_features)
batch_norm = BatchNormalization()(dense1)
dropout = Dropout(0.1)(batch_norm)

# Output layer for multiclass classification
output = Dense(num_classes, activation='softmax')(dropout)


model = Model(inputs=[text_input, emoji_input, emoji_score_input], outputs=output)


model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-4), loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.summary()

#Early stopping to complete the process faster
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

model.fit([X_train_padded, X_train_emoji_embeddings, X_train_emoji_scores], y_train,
          validation_data=([X_test_padded, X_test_emoji_embeddings, X_test_emoji_scores], y_test),
          epochs=20, batch_size=32, callbacks=[early_stopping])

dl_preds = model.predict([X_test_padded, X_test_emoji_embeddings, X_test_emoji_scores])
dl_preds = np.argmax(dl_preds, axis=1)

print(f"DL Model Accuracy: {accuracy_score(y_test, dl_preds)}, F1: {f1_score(y_test, dl_preds, average='weighted')}")


100%|██████████| 458M/458M [00:07<00:00, 62.1MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/rtatman/glove-global-vectors-for-word-representation/versions/1
316006                trying win cash know get squat though
316007    lesson learned tennis table tennis badminton u...
316008                 picking photos photo shoot wednesday
316009                                  drown naked revenge
316010        new blog post working miniminscom weight loss
Name: text_with_emojis_replaced, dtype: object
0         [0.7759947, 0.91619855, -0.03053476, 1.7538463...
1         [0.7759947, 0.91619855, -0.03053476, 1.7538463...
2         [0.7759947, 0.91619855, -0.03053476, 1.7538463...
3         [0.7759947, 0.91619855, -0.03053476, 1.7538463...
4         [0.7759947, 0.91619855, -0.03053476, 1.7538463...
                                ...                        
316006    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
316007    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
316008    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ text_input (InputLayer)   │ (None, 100)            │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ embedding (Embedding)     │ (None, 100, 200)       │      1,000,200 │ text_input[0][0]       │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ bidirectional             │ (None, 100, 256)       │        336,896 │ embedding[0][0]        │
│ (Bidirectional)           │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv1d (Conv1D)           │ (None, 96, 64)         │         81,984 │ bidirectional[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling1d             │ (None, 48, 64)         │              0 │ conv1d[0][0]           │
│ (MaxPooling1D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attention (Attention)     │ (None, 48, 64)         │              0 │ max_pooling1d[0][0],   │
│                           │                        │                │ max_pooling1d[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ flatten (Flatten)         │ (None, 3072)           │              0 │ attention[0][0]        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ emoji_input (InputLayer)  │ (None, 300)            │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ emoji_score_input         │ (None, 1)              │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ concatenate (Concatenate) │ (None, 3373)           │              0 │ flatten[0][0],         │
│                           │                        │                │ emoji_input[0][0],     │
│                           │                        │                │ emoji_score_input[0][… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense (Dense)             │ (None, 128)            │        431,872 │ concatenate[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization       │ (None, 128)            │            512 │ dense[0][0]            │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dropout (Dropout)         │ (None, 128)            │              0 │ batch_normalization[0… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_1 (Dense)           │ (None, 4)              │            516 │ dropout[0][0]          │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 1,851,980 (7.06 MB)

 Trainable params: 851,524 (3.25 MB)

 Non-trainable params: 1,000,456 (3.82 MB)

Epoch 1/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 126s 15ms/step - accuracy: 0.6907 - loss: 1.2015 - val_accuracy: 0.7294 - val_loss: 0.5892
Epoch 2/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 137s 15ms/step - accuracy: 0.7414 - loss: 0.5682 - val_accuracy: 0.7510 - val_loss: 0.5470
Epoch 3/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 141s 15ms/step - accuracy: 0.7522 - loss: 0.5451 - val_accuracy: 0.7568 - val_loss: 0.5409
Epoch 4/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 143s 15ms/step - accuracy: 0.7615 - loss: 0.5299 - val_accuracy: 0.7325 - val_loss: 0.5707
Epoch 5/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 142s 15ms/step - accuracy: 0.7676 - loss: 0.5194 - val_accuracy: 0.7629 - val_loss: 0.5282
Epoch 6/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 142s 15ms/step - accuracy: 0.7738 - loss: 0.5107 - val_accuracy: 0.7371 - val_loss: 0.5664
Epoch 7/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 141s 15ms/step - accuracy: 0.7790 - loss: 0.5007 - val_accuracy: 0.7613 - val_loss: 0.5320
Epoch 8/20
7901/7901 ━━━━━━━━━━━━━━━━━━━━ 143s 15ms/step - accuracy: 

In [ ]:
print(dl_preds[0:100])

[3 0 0 0 0 0 3 3 2 3 0 3 0 0 3 0 3 3 3 0 3 3 0 0 3 0 0 0 0 3 0 0 0 3 0 0 0
 0 0 3 3 0 3 0 0 3 0 3 3 0 3 3 3 0 0 0 3 3 0 3 3 3 3 0 3 3 0 0 0 0 3 3 0 3
 0 0 0 0 3 3 0 3 3 0 3 0 0 0 3 0 3 0 0 0 0 0 0 0 3 3]


In [ ]:
# import numpy as np
# import tensorflow as tf
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.model_selection import train_test_split
# from tensorflow.keras.preprocessing.sequence import pad_sequences
# from tensorflow.keras.layers import Embedding, LSTM, Bidirectional, Conv1D, MaxPooling1D, Dense, Dropout, Flatten, Concatenate, Input
# from tensorflow.keras.models import Model
# from sklearn.metrics import accuracy_score, f1_score

# # Function to preprocess input text and predict sentiment
# def predict_sentiment(input_text):

#     # Preprocess the input text similar to training
#     processed_text = preprocess_text(input_text)
#     processed_text = replace_emojis(processed_text)
#     padded_text = pad_sequences([vectorizer.transform([processed_text]).toarray()[0]],maxlen=100)

#     # Get emoji score and embeddings for the input text
#     emoji_score = calculate_emoji_score(processed_text)
#     emoji_embedding = get_emoji_embeddings(processed_text)

#     # Predict sentiment
#     prediction = model.predict([padded_text, np.array([emoji_embedding]), np.array([emoji_score])])
#     sentiment_label = np.argmax(prediction)
#     print(np.array([emoji_embedding]), np.array([emoji_score]))

#     # Mapping numerical prediction to sentiment label
#     sentiment_map = {0: "Sad", 1: "Happy", 2: "Angry", 3:"Love"}
#     sentiment_text = sentiment_map[sentiment_label]
#     print(f"Sentiment: {sentiment_text}")

# #inText = input("Enter a sentence with emoji or not:")
# predict_sentiment(" happy")
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.sequence import pad_sequences


def predict_sentiment(input_text):

    processed_text = preprocess_text(input_text)
    processed_text = replace_emojis(processed_text)
    tokenized_text = tokenizer.texts_to_sequences([processed_text])
    padded_text = pad_sequences(tokenized_text, maxlen=max_seq_length, padding='post')
    emoji_score = calculate_emoji_score(processed_text)
    emoji_embedding = get_emoji_embeddings(processed_text)
    emoji_embedding = np.reshape(emoji_embedding, (1, 300))
    emoji_score = np.reshape(emoji_score, (1, 1))
    prediction = model.predict([padded_text, emoji_embedding, emoji_score])
    print("Prediction probabilities:", prediction)
    sentiment_label = np.argmax(prediction)
    sentiment_map = {0: "Negative", 1: "Positive", 2: "Negative", 3: "Positive"}
    sentiment_text = sentiment_map[sentiment_label]
    print(f"Sentiment: {sentiment_text}")


test_sentences = [
    "I really didn't enjoy the book I read. 😒",
    "I need to finish my report by Friday.",
    "I’m excited about the concert, but I hope the weather holds up. 🎶🌧️",
    "I'm so grateful for my supportive friends. 🙏",
    "The traffic today was unbearable! 😤",
    "My friend is coming over later today.",
    "I enjoyed the movie, but it was a bit too long for my taste. 😅",
    "This is the best cake I've ever tasted! 🍰",
    "I feel so tired and unmotivated today. 😩",
    "The weather is nice but a bit chilly.",
    "The service at that restaurant was disappointing. 😠",
    "I plan to go for a run after work.",
    "I can't stand how rude people can be sometimes. 👎",
    "Just another day in the office.",
    "I appreciate the help, but I wish it had been offered sooner. 🤔"
]

for sentence in test_sentences:
    predict_sentiment(sentence)






1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
Prediction probabilities: [[0.08034454 0.00118669 0.01307686 0.9053919 ]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
Prediction probabilities: [[0.21834981 0.00278788 0.00835688 0.7705054 ]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step
Prediction probabilities: [[0.8530772  0.0016243  0.00343605 0.14186253]]
Sentiment: Negative
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
Prediction probabilities: [[0.24439757 0.00356866 0.01174893 0.7402849 ]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
Prediction probabilities: [[0.48973805 0.00301614 0.00817899 0.4990668 ]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 37ms/step
Prediction probabilities: [[0.42446664 0.00410646 0.00985809 0.56156886]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step
Prediction probabilities: [[0.24848492 0.00232698 0.00658084 0.74260724]]
Sentiment: Positive
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
Prediction probabilities: [[0.5241241 

In [ ]:
pip install lime


In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.sequence import pad_sequences
from lime.lime_text import LimeTextExplainer

# Function to preprocess and predict sentiment
def predict_sentiment(input_text):
    processed_text = preprocess_text(input_text)
    processed_text = replace_emojis(processed_text)
    tokenized_text = tokenizer.texts_to_sequences([processed_text])
    padded_text = pad_sequences(tokenized_text, maxlen=max_seq_length, padding='post')
    emoji_score = calculate_emoji_score(processed_text)
    emoji_embedding = get_emoji_embeddings(processed_text)
    emoji_embedding = np.reshape(emoji_embedding, (1, 300))
    emoji_score = np.reshape(emoji_score, (1, 1))

    prediction = model.predict([padded_text, emoji_embedding, emoji_score])
    return prediction  # Returning the prediction probabilities

# Function that LIME will use to get prediction probabilities from input text
def lime_predict(texts):
    # For LIME, we need to process the text and get predictions
    predictions = []
    for text in texts:
        pred = predict_sentiment(text)
        predictions.append(pred[0])  # Append the predicted probabilities for each class
    return np.array(predictions)

# Initialize the LIME text explainer
explainer = LimeTextExplainer(class_names=['Negative', 'Positive'])

# Test sentences from your sample
test_sentences = [
    "I really didn't enjoy the book I read. 😒",
    "I need to finish my report by Friday.",
    "I’m excited about the concert, but I hope the weather holds up. 🎶🌧️",
    "I'm so grateful for my supportive friends. 🙏",
    "The traffic today was unbearable! 😤",
    "My friend is coming over later today.",
    "I enjoyed the movie, but it was a bit too long for my taste. 😅",
    "This is the best cake I've ever tasted! 🍰",
    "I feel so tired and unmotivated today. 😩",
    "The weather is nice but a bit chilly.",
    "The service at that restaurant was disappointing. 😠",
    "I plan to go for a run after work.",
    "I can't stand how rude people can be sometimes. 👎",
    "Just another day in the office.",
    "I appreciate the help, but I wish it had been offered sooner. 🤔"
]

# Iterate over each sentence and generate LIME explanations
for sentence in test_sentences:
    # Explain the prediction for each sentence
    exp = explainer.explain_instance(sentence, lime_predict, num_features=10)

    # Display explanation in notebook or terminal
    exp.show_in_notebook(text=True)  # If running in a Jupyter notebook
    # Alternatively, you can save the explanations to HTML files
    exp.save_to_file(f'lime_explanation_{sentence[:10]}.html')


Example output:

Explanation for sentence: "I really didn't enjoy the book I read. 😒"
Positive prediction contribution: "really didn't enjoy" (strongly negative sentiment)
Negative prediction contribution: "book" (neutral/weak negative)
Emoji contribution: "😒" (strong negative sentiment)
Final Sentiment: Negative (0.85 confidence)
